### PR TITLE
[NFC][SROA] Remove implementation details from SROA header

### PR DESCRIPTION
**Pull request by brunodf-gf**
This moves the SROA implementation from SROAPass into a separate SROA class that is defined in the cpp file, and reduces the SROAPass class to a thin NewPM wrapper. This allows to remove all implementation details from the SROA header, and the SROALegacyPass can wrap the SROA class instead of the NewPM SROAPass.

The trigger for this change is a GCC warning about visibility of implementation details in the SROA header after D138238. Credits to Nikita Popov for suggesting this reorganization.

### CODE REVIEW

**Comment by brunodf-snps** [comment]
In this version, the passes (NewPM and legacy) create a new SROA object for each function, so the SROA data structures are reallocated in between functions. If the allocation has an impact on compile-time, the pass objects could keep an SROA member (the NewPM SROAPass with a pimpl pattern).

File: llvm/lib/Transforms/Scalar/SROA.cpp [hunk 1]
```diff
@@ -123,6 +125,140 @@ static cl::opt<bool> SROASkipMem2Reg("sroa-skip-mem2reg", cl::init(false),
...
+       SROAOptions PreserveCFG_)
+      : C(C), DTU(DTU), AC(AC),
+        PreserveCFG(PreserveCFG_ == SROAOptions::PreserveCFG) {}
+
+  /// Helper used by both the SROAPass and by the legacy pass.
+  std::pair<bool /*Changed*/, bool /*CFGChanged*/> runSROA(Function &F);
+
+private:
+  friend class AllocaSliceRewriter;
+  friend class SROALegacyPass;
```
Comment (nikic) [inline llvm/lib/Transforms/Scalar/SROA.cpp:249]: Why do we need this friend class?

Comment (brunodf-snps) [reply llvm/lib/Transforms/Scalar/SROA.cpp]: You are right, it is a leftover from `SROALegacyPass` invoking `runImpl`, but the new `runSROA` method is public. I've removed the friend declaration now.

**Review by nikic — APPROVED** [review]
LGTM

**Review by XJTUDing — COMMENTED** [review]
why no test case for this patch ?

**Pull request merged**